XDG base directory specification support

ZyX-I opened the pull request:
This is the replacement for #3198, that fully implements #78, including documentation update, as requested in https://github.com/neovim/neovim/pull/3198#issuecomment-148904123.

I have not touched pi_*.txt, os_*.txt, vi_diff.txt, todo.txt, various.txt files.

Note: since it was not discussed, I assumed that `~/.gvimrc` is replaced with `$XDG_CONFIG_HOME/nvim/ginit.vim`. Unless it is going away though.

ZyX-I committed:
os/env: Add functions that can iterate over colon-separated variables

ZyX-I committed:
option: Use different default value for &runtimepath

ZyX-I committed:
option: Add /site subdirectory to data directories

ZyX-I committed:
option: Move all data directories after all config directories

ZyX-I committed:
option: Prepare for all environment variables being NULL.

ZyX-I committed:
option: Append with escaped commas

jck committed:
option,main: Partial support of XDG base directory specification

- Add functions that are able to query XDG.
- Replace defaults for
  - &runtimepath. Does not follow #78.
  - &viewdir.
  - &undodir.
  - &directory.
  - &backupdir. Does not follow #78.
  - vimrc location.
- Remove user vimrc file line from :version message.

ZyX-I committed:
stdpaths: Fix some issues, specifically

- Remove/add includes
- Do not pretend that returns are const
- Add function attributes
- Allocate memory always in get_xdg

ZyX-I committed:
stdpaths: Remove useless functions get_user_*_dir

ZyX-I committed:
stdpaths: Rename export functions so that they have common prefix

ZyX-I committed:
stdpaths: Export get_xdg function (renamed) and use it for runtimepath

ZyX-I committed:
option: Remove some memory leaks

ZyX-I committed:
option: Use proper printexpr

ZyX-I committed:
stdpaths,main: Remove all remaining memory leaks

ZyX-I committed:
shada: Move shada file to a new location

ZyX-I committed:
stdpaths: Give proper error message in case directory creation failed

ZyX-I committed:
stdpaths: Do NOT create data directories

This is none of option.c business to create *possibly unneeded* **default**
directories **before** user specified where he actually wants to place the
files.

ZyX-I committed:
documentation: Update documentation

Note about ~/.local/share/nvim/site used in one usr_\* file: this one talks
about user-local installation of third-party plugins, and
~/.local/share/nvim/site is the proper place for them. Most other files talk
about user own configuration and this is ~/.config.

ZyX-I commented:
Most of documentation changes is replacing `.vimrc` (with dot) with `vimrc` (without dot).

ZyX-I commented:
Also many removes regarding windows (as I have no idea what should be written there).

ZyX-I committed:
option: Move macros to functions, use PATHSEP in place of /

ZyX-I committed:
option: Do not add unneeded path separator, protect against zero len

ZyX-I commented:
https://github.com/neovim/neovim/commit/379513671849006db68bf213b5eb23175dca23bb hanged on unknown test. If https://github.com/neovim/neovim/commit/3796471038fd9e3ed1ab3d20c8ca000fd73f1d78 also hangs I am considering #3427 a blocker (instead of using temporary commit that sets gtest output).

BTW, https://github.com/ZyX-I/neovim/commit/dcc02b23e6b94a9e7d389410acd051622614938c should not be part of this PR, but I do not know whether this should be part of a separate PR, or of something windows-related.

ZyX-I committed:
option: Remove new trailing spaces

ZyX-I committed:
main,os/env: Fix lint errors

justinmk commented:
https://github.com/ZyX-I/neovim/commit/dcc02b23e6b94a9e7d389410acd051622614938c is ok for this PR.

ZyX-I commented:
This one also hangs. Hang does not happen locally, thus I am waiting for resolution with #3427.

ZyX-I commented:
By the way, not that this PR originally has `&backupdir` value different from #78: specifically `.,` prefix is absent (no backups inside current directory).

fmoralesc commented on src/nvim/os/stdpaths.c:
```diff
@@ -0,0 +1,101 @@
+#ifdef WIN32
+  // Windows
+#elif APPLE
+  // Apple (this includes iOS, which we might need to handle differently)
```
I think we decided to use the same XDG paths as in other *nixes for OSX (see https://github.com/neovim/neovim/pull/3198#discussion_r41556908 and following)

justinmk replied:
Yes :+1:  Treat OSX as unix. Haters gonna hate.

justinmk commented:
> &backupdir value different from #78: specifically ., prefix is absent (no backups inside current directory).

:+1: One of the major advantages of adopting XDG is that we have standard directories to persist swap/undo/backup. We should default to using those and consider enabling swap & persistent undo by default.

ZyX-I commented:
Also it would be good to merge #1909 before this: to place socket in $XDG_RUNTIME_DIR I in any case need a single entry point.

ZyX-I committed:
stdpaths: Add documentation

justinmk commented on src/nvim/version.c:
```diff
@@ -1067,21 +1067,6 @@ void list_version(void)
-  version_msg(USR_VIMRC_FILE3);
-  version_msg("\"\n");
-#endif  // ifdef USR_VIMRC_FILE3
 #ifdef USR_EXRC_FILE
```
I do not think it makes sense to support `USR_EXRC_FILE` or `USR_EXRC_FILE2` if we are removing vimrc already.

justinmk commented on src/nvim/shada.c:
```diff
@@ -1608,11 +1622,11 @@ static char *shada_filename(const char *file)
 #endif
-          file =  SHADA_FILE;
+          file =  shada_get_default_file();
 #ifdef SHADA_FILE2
```
When would `SHADA_FILE2` be defined? Why is it optional?

ZyX-I replied:
I have no idea. Inherited from viminfo. Maybe for windows `.`/`_` support, in which case it is not needed (no leading dots in file name anymore).

justinmk replied:
Even if so the `_` prefix is not need in modern Windows.

ZyX-I commented:
I have a reason to want `.,` default there: until we have support for `//` like `&directory` there will be overlaps.

fmoralesc commented on src/nvim/os/stdpaths.c:
```diff
@@ -0,0 +1,101 @@
+static const char *const xdg_defaults[] = {
+  // Windows, Apple stuff are just shims right now
+#ifdef WIN32
+  // Windows
```
If we leave this empty for now, what would it mean for people who are building/using nvim in Windows already?

ZyX-I replied:
17.10.2015, 21:41, "Felipe Morales" <notifications@github.com>:
> In src/nvim/os/stdpaths.c:
>
>> +static const char *xdg_env_vars[] = { > + [kXDGConfigHome] = "XDG_CONFIG_HOME", > + [kXDGDataHome] = "XDG_DATA_HOME", > + [kXDGCacheHome] = "XDG_CACHE_HOME", > + [kXDGRuntimeDir] = "XDG_RUNTIME_DIR", > + [kXDGConfigDirs] = "XDG_CONFIG_DIRS", > + [kXDGDataDirs] = "XDG_DATA_DIRS", > +}; > + > +/// Defaults for XDGVarType values > +/// > +/// Used in case environment variables contain nothing. Need to be expanded. > +static const char *const xdg_defaults[] = { > + // Windows, Apple stuff are just shims right now > +#ifdef WIN32 > + // Windows
>
> If we leave this empty for now, what would it mean for people who are building/using nvim in Windows already?

That this will fail with syntax error. #endif is wrongly placed.

If it is correctly placed it should SEGV.

> —
> Reply to this email directly or view it on GitHub.

justinmk replied:
We can stub it out with the examples from https://github.com/ActiveState/appdirs/blob/master/appdirs.py .

ZyX-I replied:
@justinmk Even assuming that `C:\\Users\\user\\` from examples is `~\\` (I can’t reliably use `$USER` even on \*nix) I have no idea what is `AppAuthor`. Also what exactly is `AppName`: nvim/Neovim/neovim?

fmoralesc replied:
@ZyX-I  `AppAuthor` in this case should be `Neovim`, and the same for `AppName`. Supposedly, you could use the `%AppData%` environment variable, see http://blogs.msdn.com/b/patricka/archive/2010/03/18/where-should-i-store-my-data-and-configuration-files-if-i-target-multiple-os-versions.aspx

justinmk replied:
@ZyX-I @fmoralesc I am working on a code sample, give me a minute.

ZyX-I committed:
option: Add `//` to the end of default `&directory`

ZyX-I committed:
main,version: Remove USR_EXRC_FILE*

ZyX-I committed:
stdpaths: Remove Apple defaults, use \*nix ones instead

Pyrohh commented on runtime/doc/autocmd.txt:
```diff
@@ -994,7 +994,7 @@ expect.
 	:autocmd BufRead $VIMRUNTIME/doc/*.txt  set expandtab
 And ~ can be used for the home directory (if $HOME is defined): >
-	:autocmd BufWritePost ~/.vimrc   so ~/.vimrc
+	:autocmd BufWritePost ~/.config/nvim/init.vim   so <afile>
```
Regarding all these references to `~/.config/...`: shouldn't these instead say `$XDG_CONFIG_HOME/...`?

ZyX-I replied:
17.10.2015, 22:01, "Michael Reed" <notifications@github.com>:
> In runtime/doc/autocmd.txt:
>
>> @@ -994,7 +994,7 @@ expect. > Environment variables can be used in a pattern: > > :autocmd BufRead $VIMRUNTIME/doc/*.txt set expandtab > And ~ can be used for the home directory (if $HOME is defined): > > - :autocmd BufWritePost ~/.vimrc so ~/.vimrc > + :autocmd BufWritePost ~/.config/nvim/init.vim so <afile>
>
> Regarding all these references to ~/.config/...: shouldn't these instead say $XDG_CONFIG_HOME/...?

Not all of them: $XDG_CONFIG_HOME is not guaranteed to be defined since PR with approach which adds support for XDG_* vars to vim_getenv was rejected.

Most of other stay as-is because I am not fond of explaining WTF is $XDG_CONFIG_HOME and what is its default. I guess we need another section somewhere in documentation (options.txt?) with explanations.

>
> —
> Reply to this email directly or view it on GitHub.

Pyrohh replied:
> I guess we need another section somewhere in documentation (options.txt?) with explanations.
I think that would be a good idea. It's not too complicated, so I think it should be doable in a paragraph or two, at least as a high level overview.

justinmk replied:
I don't think it's necessary to be so specific. It is understood that `~/.config` means "what ~/.vim used to mean". And for new users, also, it is easier to understand what `~/.config` than an env var from the linux desktop standard.

Pyrohh replied:
Eh, probably. Let's not complicate it.

Pyrohh commented on runtime/doc/starting.txt:
```diff
@@ -373,24 +373,24 @@ accordingly.  Vim proceeds in this order:
-		Unix		$HOME/.vimrc or $HOME/.vim/vimrc
-		MS-Windows	$HOME/_vimrc, $HOME/vimfiles/vimrc
-				or $VIM/_vimrc
+		Unix		$XDG_CONFIG_HOME/.config/nvim/init.vim
```
shouldn't this be `Unix		$XDG_CONFIG_HOME/nvim/init.vim`

Pyrohh commented on runtime/doc/starting.txt:
```diff
@@ -373,24 +373,24 @@ accordingly.  Vim proceeds in this order:
-	$HOME/.vim/ directory ($HOME/vimfiles/ for MS-Windows). That makes it
-	easy to copy it to another system.
+	RECOMMENDATION: Put all your Vim configuration stuff in the 
+	$HOME/.config/nvim/ directory. That makes it easy to copy it to 
```
`$XDG_CONFIG_HOME` was just  introduced above, so I think it should be used here instead of `$HOME/.config`.

Pyrohh commented on runtime/doc/filetype.txt:
```diff
@@ -224,7 +224,7 @@ D. If your filetype can only be detected by inspecting the contents of the
    1. Create your user runtime directory.  You would normally use the first
       item of the 'runtimepath' option.  Example for Unix: >
-	:!mkdir ~/.vim
+	:!mkdir ~/.config/nvim
```
`mkdir -p`?

Pyrohh commented on runtime/doc/filetype.txt:
```diff
@@ -196,7 +196,7 @@ B. If you want to detect your file after the default file type checks.
    1. Create your user runtime directory.  You would normally use the first
       item of the 'runtimepath' option.  Example for Unix: >
-	:!mkdir ~/.vim
+	:!mkdir ~/.config/nvim
```
`mkdir -p`?

Pyrohh commented on runtime/doc/filetype.txt:
```diff
@@ -162,8 +162,8 @@ A. If you want to overrule all default file type checks.
-	:!mkdir ~/.vim
-	:!mkdir ~/.vim/ftdetect
+	:!mkdir ~/.config/nvim
+	:!mkdir ~/.config/nvim/ftdetect
```
`mkdir -p`?

ZyX-I committed:
shada: Remove SHADA_FILE2

justinmk commented:
XDG defaults for Windows:

```diff
+  [kXDGConfigHome] = "~\\$CSIDL_LOCAL_APPDATA\\nvim",
+  [kXDGDataHome]   = "~\\$CSIDL_LOCAL_APPDATA\\nvim",  //same as kXDGConfigHome
+  [kXDGCacheHome]  = "~\\$CSIDL_LOCAL_APPDATA\\nvim\\cache",
+  [kXDGRuntimeDir] = "",  //I don't know what this is for
+  [kXDGConfigDirs] = "$CSIDL_COMMON_APPDATA\\nvim\\nvim",
+  [kXDGDataDirs] = "$CSIDL_COMMON_APPDATA\\nvim\\nvim",  //same as kXDGConfigDirs
```

Note that this does not switch on "roaming" but that is out of scope for now. This should be good enough.

ZyX-I commented:
@justinmk Try `echo %CSIDL_LOCAL_APPDATA%` in `wineconsole cmd.exe`. It will echo `%CSIDL_LOCAL_APPDATA%`. For `%PATH%` it works fine.

ZyX-I commented:
That means that there is no such environment variable we can rely upon.

ZyX-I commented:
cmd.exe in Windows 8.1 in VirtualBox makes no difference.

fmoralesc commented:
@ZyX-I What do `%APPDATA%` and

ZyX-I commented:
`%APPDATA` works (C:\\users\\zyx\\Application Data in wine). Not sure what was supposed to be written after “and”.

Pyrohh commented:
Given that a user vimrc could be in `~/.nvimrc` or `~/.nvim/nvimrc` by default, referring to it as `.n?vimrc` doesn't seem preferable regardless of the XDG spec being implemented or not.

Given that, I think most of the documentation touched in this PR can be extracted to a separate PR. I'll go ahead and do that if others agree.

ZyX-I commented:
`LOCALAPPDATA` also works in wine (adds Local Settings after zyx).

ZyX-I commented:
And I got “@ZyX-I What do %APPDATA% and” for https://github.com/neovim/neovim/pull/3470#issuecomment-148946186 in email, thus “Not sure what was supposed to be written after “and”.”.

ZyX-I commented:
@Pyrohh This would be a number of merge conflicts, especially given that some `.vimrc` occurences were replaced with `init.vim` and not `vimrc`.

oni-link commented on src/nvim/os/stdpaths.c:
```diff
@@ -0,0 +1,102 @@
+  if (trailing_pathseps) {
+    const size_t len = strlen(ret);
+    ret = xrealloc(ret, len + trailing_pathseps + 1);
+    memset(ret + len, PATHSEP, trailing_pathseps);
```
Compiler complains about `PATHSEP` being undeclared.

oni-link commented on src/nvim/os/stdpaths.c:
```diff
@@ -0,0 +1,102 @@
+    const size_t len = strlen(ret);
+    ret = xrealloc(ret, len + trailing_pathseps + 1);
+    memset(ret + len, PATHSEP, trailing_pathseps);
+    ret[len + trailing_pathseps] = NUL;
```
Compiler complains about `NUL` being undeclared.

fmoralesc commented:
@ZyX-I Sorry, I pressed the Enter key by accident. I was asking for the value of `%LOCALAPPDATA%` (`%APPDATA%` is roaming, `%LOCALAPPDATA%` isn't).

Pyrohh commented:
@ZyX-I 

> especially given that some .vimrc occurences were replaced with init.vim and not vimrc

Those were the ones I planned on not touching. The reason I suggested it is because many--if not most--of the doc changes seem generally applicable. This PR is huge as it is, so slimming it down would help reviewing.

I was also thinking the USR_EXRC_* stuff could be extracted out. I remember @justinmk mentioned that a while ago.

ZyX-I committed:
stdpaths: Add missing include

ZyX-I committed:
functests: Fix 078_swapfile_recover test

justinmk commented:
> LOCALAPPDATA also works in wine (adds Local Settings after zyx).

@ZyX-I You are right that `%CSIDL_LOCAL_APPDATA%` does not work in cmd.exe. For that we may need to use win32 API, e.g. `SHGetFolderPath`. Just use `LOCALAPPDATA` env var for now.

Do _not_ use `APPDATA`, it resolves to `~/AppData/Roaming` if current environment is roaming. `LOCALAPPDATA` is least-risk for now, until we add enhancement for roaming and non-roaming cases.

oni-link commented on src/nvim/os/stdpaths.c:
```diff
@@ -0,0 +1,103 @@
+
+  const char *const env_val = os_getenv(env);
+  char *ret = NULL;
+  if (ret != NULL) {
```
`env_val` instead of `ret`?

ZyX-I commented:
@justinmk I do not like the idea of XDG_CONFIG_HOME and XDG_DATA_HOME pointing to the same locations: this will turn this directory into a junk, which will not cause problems with double loading of the plugins solely because I decided to use `site` subdirectory for data dirs.

oni-link commented on src/nvim/os/stdpaths.c:
```diff
@@ -0,0 +1,103 @@
+char *stdpaths_user_conf_subpath(const char *fname)
+  FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
+{
+  return concat_fnames_realloc(get_xdg_home(kXDGConfigHome), fname, true);
```
At the moment `get_xdg_home()` could return `NULL` (empty `xdg_defaults` array for Windows), but `concat_fnames_realloc()` needs a not `NULL` argument.

ZyX-I replied:
Empty xdg_defaults is a crash, not NULL.

ZyX-I replied:
Because it expands to `xdg_defaults[] = {}` which has zero length. Or compilation error (clang thinks this is GNU empty initializer extension).

oni-link replied:
Should the entries for Windows set to an empty string until we have something to put into them or is this intentional?

ZyX-I replied:
@oni-link What to put there is currently being discussed (and added in the most recent commit). Just wait.

justinmk commented:
@ZyX-I That's fine, we can choose whatever subdirectories we want in that case, I was only using the ActiveState examples. The main takeaways are that we can use `$LOCALAPPDATA` for user-local XDG dirs, and `$ProgramData` for system-local dirs.

ZyX-I commented:
@justinmk Are you sure `$ProgramData` is defined? I do not see it in wine+cmd.exe.

Also it is completely pointless to define CONFIG_DIRS and DATA_DIRS in the same location as CONFIG_HOME and DATA_HOME. So I am using NULL.

justinmk commented:
@ZyX-I `$ProgramData` is defined on all Windows 7 ("Professional") machines I run. Not sure about "Home" edition but I know that this is also defined in Vista Professional.

ZyX-I commented:
@justinmk I do not see this in XP+cmd.exe. Saw in 8.1. Wine is also emulating XP.

ZyX-I committed:
stdpaths: Add Windows-specific directories

justinmk commented:
@ZyX-I The ActiveState code mentions that they don't actually use XDG_DATA_DIRS nor XDG_CONFIG_DIRS code on Windows:

https://github.com/ActiveState/appdirs/blob/master/appdirs.py#L129

https://github.com/ActiveState/appdirs/blob/master/appdirs.py#L234

So yeah, NULL is fine.

ZyX-I commented:
@justinmk Current code is going to use environment variables in any case. NULL just means “no default”.

ZyX-I committed:
option: Silence “may be used unitialized” errors

ZyX-I committed:
stdpaths: Remove outdated comment

oni-link commented on src/nvim/version.c:
```diff
@@ -1067,31 +1067,6 @@ void list_version(void)
-  version_msg(_("  2nd user exrc file: \""));
-  version_msg(USR_EXRC_FILE2);
-  version_msg("\"\n");
-#endif  // ifdef USR_EXRC_FILE2
```
New locations for user files?

ZyX-I replied:
`:version` message is telling about neovim version, compilation options, who compiled it. Now vimrc file location does not depend on compilation options so it has absolutely nothing to do here.

justinmk replied:
Users may still like to see the paths listed in :version.

ZyX-I replied:
This breaks the purpose of `:version`. And `$MYVIMRC` is still there. Also vimrc file name placed here means that without `-u` Neovim will *always* source *exactly this* vimrc which

1. Is useless if it is written as `$XDG_CONFIG_HOME/.config/init.vim`. (BTW, I have not yet implemented #78/1/ii which will make this variant look even more strange.)
2. Contains even more information which has nothing to do with compilation options if I also provide $XDG_CONFIG_HOME (and $XDG_CONFIG_DIRS) values in addition to their defaults.
3. Is generally incorrect if I perform the expansion before providing any information, and also will contain lots of unrelated information with 1/ii.
4. Is useless for the purpose of determining “where I need to put init.vim file” in case I provide only the effective file from all the variants with 1/ii (still useful if 1/ii is not implemented).

justinmk replied:
Good points.

ZyX-I commented:
Travis for some reason is not going to build anything in a meantime, and quickbuild is as usual useless for determining “what hanged”. Nothing hangs or fails locally. Though mf_hash_grow, “not hanging” grep and pasting “large chunks of input” tests take rather long time, as well as my shada “different times” jumps test.

Total 32 seconds (busted “32 seconds” are zsh “203,37s user 37,16s system 107% cpu 3:44,56 total”) for functional tests, half a second (23,66s user 1,65s system 141% cpu 17,844 total) for unit tests.

ZyX-I commented:
@jszakmeister By the way, unit+functional tests are now taking greater then four minutes of “real” time on my system (zsh by default reports, in order: “CPU seconds spent in user mode”, `user`, “CPU seconds spent in kernel mode”, `system`, “The CPU percentage, computed as 100*(%U+%S)/%E”, `cpu`, “Elapsed time in seconds”, `total`). I am wondering what is QB measuring since in http://neovim-qb.szakmeister.net/build/3300 they were recorded as completed in less then a minute.

This PR should not add any time to the test build time.

ZyX-I commented:
Though I think there needs to be a way to query effective values of `XDG_\*` variables. Not necessary `vim_getenv` (and `$XDG_*` in VimL expressions), but, at least, some function in `eval.c`.

By the way, what to do with XDG_RUNTIME_DIR default? AFAIR last discussion was around choosing between `/tmp` and `~/.run`. There is also a variant “no default, no socket until user comes up with something”: if I understood right this is not going to harm systemd users (systemd sets this variable), unless they do something strange (this is still possible to unset/override it). http://serverfault.com/questions/388840/good-default-for-xdg-runtime-dir suggests something like `/tmp/app-$USER` (I would use `/tmp/run-$USER`: in any case NeoVim itself will be using `/tmp/run-$USER/nvim`), but this is tricky: `$USER` is not guaranteed to be defined, also `$TMPDIR` for PAM users.

Now I think most logical is to write a function which will be the expansion of `${TMPDIR:-/tmp}/run-$USER[.$I]` written in C, where `$USER` is obtained using `os_get_user_name`. `.$I` suffix is appended in case `${TMPDIR:-/tmp}/run-$USER` exists, but does not satisfy security restrictions. It would also be logical to check all `${TMPDIR:-/tmp}/run-$USER{,.[1-9][0-9]#}` and get one with greatest `$I` (without suffix it is considered to be I=0) which matches security restrictions, but since we need XDG_RUNTIME_DIR only for creating sockets inside, and do not really care about all Neovim instances writing to the same directory it is not needed I think.

oni-link commented on src/nvim/option.c:
```diff
@@ -301,6 +301,256 @@ static char *(p_cot_values[]) = {"menu", "menuone", "longest", "preview",
 #endif
 
+/// Count commas in the given string
+static size_t count_commas(const char *const s, size_t len)
```
`memcnt(s, ',', len)` in `memory.c` does the same thing.

ZyX-I committed:
option: Use memcnt for counting commas

justinmk commented:
Build cancel & restart did not help. I have emailed travis support, awaiting reply.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -6762,8 +6776,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 						*'viewdir'* *'vdir'*
-'viewdir' 'vdir'	string	(default for MS-DOS and Win32: "$VIM/vimfiles/view",
-				 for Unix: "~/.vim/view")
+'viewdir' 'vdir'	string	(default: "$XDG_DATA_HOME/nvim/view")
```
`'directory'` doc needs a similar update.

kopischke replied:
I know we have discussed this before in #78, but I’m still wondering if we are correct in defining view session files, backups, swap files etc. as **essential** data going into `$XDG_DATA_HOME` instead of as non-essential data, which would go into `$XDG_CACHE_HOME` [according to the spec](http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html).

I have had some doubts since my last [contribution to that discussion](https://github.com/neovim/neovim/issues/78#issuecomment-39668459) (which applied to the inadvisability of using `$XDG_RUNTIME_DIR`, a directory that _must_ not survive log offs or reboots), but now that this feature is nearing release, I would be remiss in not bringing up my suspicion that we might be off base in our interpretation of the XDG spec. I know I should have brought this up earlier (especially as it has [been brought up in #78](https://github.com/neovim/neovim/issues/78#issuecomment-77776048), a fact I seem to have completely missed), but are we absolutely certain that  `$XDG_DATA_HOME` isn’t me at for things like user installed spelling dictionaries and such, while state should better go into `$XDG_CACHE_HOME`?

ZyX-I replied:
`'viewdir'` *is* essential (I would though use `$XDG_CONFIG_HOME`). As you see in that discussion I think that swap files should be in CACHE, but there is a reason to use DATA.

kopischke replied:
@ZyX-I hmm, I seem to have missed the relevant discussion about placing swap files in DATA instead of CACHE (the only thing I found was my own remark, and your comment referring it, both linked above, which do not apply, as that was about not using RUNTIME). As to `viewdir` … view session files are *important*, yes, but *essential*? They look like poster children for state to me, but that might just be me, and I agree we may be discussing an edge case considering they are expressly written on user request (which would spell DATA); OTOH, the view directory is one of the state directories in need of regular purging (which would spell CACHE; CONFIG I find an unlikely choice either way).

To resume my misgivings, I think we should aim for this:

1. `&directory` default (swap files): `$XDG_CACHE_HOME/nvim/swap//`
2. `&backupdir` default:  `.,$XDG_CACHE_HOME/nvim/backup`
3. `&undodir` default: `$XDG_DATA_HOME/nvim/undo//` (no quibbles there)
4. ShaDa default: `$XDG_DATA_HOME/nvim/shada/main.shada` (ditto)
5. `&viewdir` default:  `$XDG_DATA_HOME/nvim/view` (or `$XDG_CACHE_HOME/nvim/view`)

fmoralesc replied:
@kopischke's proposal makes sense to me, so :+1: 

I would favor using `$XDG_DATA_HOME` for `&viewdir`, because of the fact that views are user created.

kopischke replied:
@fmoralesc agreed on the last point; I revised my comment accordingly.

justinmk commented:
If I run 243572c98441399e6579b1fcc38ec17ba6609bed like this:

    VIMRUNTIME=runtime/ ./build/bin/nvim  -u NONE

Swap file error occurs for pretty much every file. Also occurs for `q:`:

    E303: Unable to open swap file for "[Command Line]", recovery impossible

justinmk commented:
Migration steps for users seem to be:

```
mv ~/.vimrc ~/.vim/vimrc                                                                                                                                   
mkdir ~/.config                                                                                                                                             
ln -s ~/.vim ~/.config/nvim                                                                                                                                 
ln -s ~/.vim/vimrc ~/.config/nvim/init.vim
```

ZyX-I commented:
@justinmk My variant of migration steps should also work, and leave vimrc where it is.

ZyX-I committed:
os/fs: Allow os_mkdir_recurse directory name to end with ///

ZyX-I committed:
memline: Automatically create swap file directory for last directory

jszakmeister commented:
Do we really want to tuck swap files into the XDG structure by default?  One downside is that when working on a machine with others, having the swapfiles present near the source or in the tmp folder is a useful indicator that someone else is messing with stuff.  If you put it into the XDG structure by default, then you lose that benefit.  It's nice to not have to worry about it out of the box, but then you guys may not be catering to sysadmins either--I can see the XDG default being better for software devs who use Neovim.

ZyX-I committed:
*: Fix linter errors

ZyX-I commented:
@jszakmeister For indicating that “someone else is messing with stuff” swap files are fine in any directory: Neovim will warn. Unless you want not a warning from Neovim, but seeing swap files by other means (e.g. `ls`).

jszakmeister commented:
@ZyX-I `$HOME` is not a good directory when two people are accessing a file as themselves (I don't have access to their home directory).

ZyX-I commented:
@jszakmeister Leaving `.` in `&directory` has a big downside of scattering garbage all over the filesystem and requiring to do something to ignore this garbage in VCS. This is *much* more common problem then two people editing the same file in a shared filesystem.

jszakmeister commented:
@ZyX-I That depends on the environment you're in. :-)

As I alluded to above, I think Neovim leans a bit more towards software developers and the like, so it's a useful change for them (and me).  Though, we should probably make some noise about the change in the default because anyone counting on the current behavior will be surprised.

ZyX-I commented:
@jszakmeister I have no idea how you can find more convenient to use shared folders with file locks (swap file used this way are file locks) and not separate repositories with usual merging methods.

In any case this is only useful if

1. People are using shared folders for editing files.
2. *All* use Vim or NeoVim.
3. Vim or NeoVim *does* use swap files.
4. Vim or NeoVim is configured to use current directory.

. Too much conditions: editing file in another editor, disabling swap files (e.g. because they tend to leave garbage) or placing them to the other location will destroy this “lock” effectiveness.

jszakmeister commented:
> @jszakmeister I have no idea how you can find more convenient to use shared folders with file locks (swap file used this way are file locks) and not separate repositories with usual merging methods.

I never said that @ZyX-I.  All I was trying to say was that for sysadmins it's a great help to get a warning about someone else editing the file--and they don't always do their tasks as root.  As software developers (myself included!), we sometimes forget about the other use-cases for our tools.  I just wanted to point it out that for sysadmins who use the tool this way, it's a significant change.  I'm not downvoting it, I'm not saying we shouldn't do it, I was merely asking the question of whether it had been thought of, and now that it has, do we want to keep the default.  You're answer is "yes."  I'm cool with that.  I just want us to recognize that the behavior is significantly different for others outside the software development circle.  That's all.

oni-link commented on src/nvim/option.c:
```diff
@@ -301,6 +301,243 @@ static char *(p_cot_values[]) = {"menu", "menuone", "longest", "preview",
+                                              const size_t single_suf_len)
+  FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_PURE
+{
+  if (val == NULL && *val) {
```
Dereferencing a null pointer. Should the test be `if (val == NULL || *val == NUL)`?

oni-link commented on src/nvim/option.c:
```diff
@@ -301,6 +301,243 @@ static char *(p_cot_values[]) = {"menu", "menuone", "longest", "preview",
+                                   const bool forward)
+  FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_RET FUNC_ATTR_NONNULL_ARG(1)
+{
+  if (val == NULL && *val) {
```
Dereferencing a null pointer. Should the test be `if (val == NULL || *val == NUL)`?

oni-link commented on src/nvim/option.c:
```diff
@@ -301,6 +301,243 @@ static char *(p_cot_values[]) = {"menu", "menuone", "longest", "preview",
+static inline size_t compute_double_colon_len(const char *const val,
+                                              const size_t common_suf_len,
+                                              const size_t single_suf_len)
+  FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_PURE
```
Called function `vim_colon_env_iter()` does not look pure.

ZyX-I replied:
@oni-link It is not pure only because it has “out” arguments. This function is pure because it only reads global memory (referenced in val argument) and reads/writes to memory allocated on its stack which is why it is called pure.

oni-link replied:
In this case it looks like only stack memory is used, but I would consider it global memory from the point of the called function: gcc pure attribute description is a bit "short/unclear", what I refer to is an extract from https://lwn.net/Articles/285332/, (bold part marked by me)

> A pure function is a function with basically no side effect. This means that pure functions return a value that is calculated based on given parameters and global memory, but cannot affect the value of any other global variable. Pure functions cannot reasonably lack a return type (i.e. have a void return type). 

and
> A counter-example of a non-pure function is the strcpy() function. This function takes two pointers as parameters. It accesses the latter to read the source string, and the former to write to the destination string. As I said, **the memory areas pointed to by the parameters are not parameters on their own, but are considered global memory** and, in that function, global memory is not only accessed for reading, but also for writing. The return value derives directly from the parameters (it is the same as the first parameter), but global memory is affected by the side effect of strcpy(), making it not pure. 

According to this "interpretation" of attribute pure a functions with "out" arguments could not be pure.

I'm always a bit worried when attribute pure is used: If the user declares a function pure, is the compiler allowed to deduce that every called function in this function is also pure?

ZyX-I replied:
I leave `compute_double_colon_len` pure because if I inlined body of impure `vim_colon_env_iter` into its body it would automatically become pure. I do not think that compiler is allowed to deduce this, and have not found this requirement in GCC documentation.

ZyX-I committed:
functests: Do not forget about -i argument

Target: make all tests run with chmod -x ~/.config/nvim ~/.local/share/nvim.

ZyX-I committed:
functests: Use . for various folder defaults in tests

ZyX-I committed:
functests: Fix tests

ZyX-I committed:
oldtests: Also set `.` as default directories for old tests

ZyX-I committed:
main: Check init.vim files also in other XDG directories

ZyX-I committed:
os/unix_defs: Rename default system vimrc file to sysinit.vim

This way all standard Vim file paths have .vim extension. VIMRC_FILE constant 
used for &exrc option was not touched.

ZyX-I committed:
documentation: Update documentation regarding init.vim location

ZyX-I commented:
About an API for script writers: I would suggest to add (in another PR) function `getxdgdirs(varname)` which accepts strings like `XDG_CONFIG_HOME` and returns either a string (for `*_HOME` and `*_DIR` XDG variables) or a list (for `*_DIRS`), empty in case of undefined. For users there is documentation and tutorials (BTW I do not remember myself changing any tutorials).

justinmk commented on src/nvim/main.c:
```diff
@@ -1801,117 +1796,121 @@ static void exe_commands(mparm_T *parmp)
+      char *vimrc = xmalloc(dir_len + sizeof(path_tail) + 1);
+      memmove(vimrc, dir, dir_len);
+      vimrc[dir_len] = PATHSEP;
+      memmove(vimrc + dir_len + 1, path_tail, sizeof(path_tail));
```
though it doesn't matter here, ARRAY_SIZE would be a "best practice" instead of sizeof.

justinmk commented:
Thank you @ZyX-I .

xu-cheng commented:
Looks like this PR introduced an backward incompatibility. It doesn't respect `~/.nvimrc` or `~/.nvim/nvimrc` as configure file anymore. Is such breaking change intentional?

```
$ nvim -v
NVIM 0.0.0-alpha+201510252238 (compiled Oct 26 2015 11:25:35)
Commit: 1ca5646bb52ec5c23b28f45bb7bc5d25cffad9b0
Build type: RelWithDebInfo
Compilation: /home/comp/chengxu/usr/bin/gcc-5 -Os -w -pipe -march=core2 -Wconversion -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/tmp/neovim20151026-26575-xo616q/build/config -I/tmp/neovim20151026-26575-xo616q/src -I/tmp/neovim20151026-26575-xo616q/deps-build/usr/include -I/tmp/neovim20151026-26575-xo616q/deps-build/usr/include -I/tmp/neovim20151026-26575-xo616q/deps-build/usr/include/luajit-2.0 -I/tmp/neovim20151026-26575-xo616q/deps-build/usr/include -I/tmp/neovim20151026-26575-xo616q/deps-build/usr/include -I/tmp/neovim20151026-26575-xo616q/deps-build/usr/include -I/tmp/neovim20151026-26575-xo616q/deps-build/usr/include -I/home/comp/chengxu/usr/include -I/tmp/neovim20151026-26575-xo616q/build/src/nvim/auto -I/tmp/neovim20151026-26575-xo616q/build/include
Compiled by chengxu@csr40.comp.hkbu.edu.hk

Optional features included (+) or not (-): +acl   +iconv    +jemalloc
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/home/comp/chengxu/usr/Cellar/neovim/HEAD/share/nvim"
```

Compared with compiling revision before this PR:
```
$ nvim -v
NVIM 0.0.0-alpha+201510231011 (compiled Oct 26 2015 11:41:09)
Commit: de4cb766ca381c09fd3f938136c1932ebf008f63
Build type: RelWithDebInfo
Compilation: /home/comp/chengxu/usr/bin/gcc-5 -Os -w -pipe -march=core2 -Wconversion -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/tmp/neovim20151026-17981-bt0ppn/build/config -I/tmp/neovim20151026-17981-bt0ppn/src -I/tmp/neovim20151026-17981-bt0ppn/deps-build/usr/include -I/tmp/neovim20151026-17981-bt0ppn/deps-build/usr/include -I/tmp/neovim20151026-17981-bt0ppn/deps-build/usr/include/luajit-2.0 -I/tmp/neovim20151026-17981-bt0ppn/deps-build/usr/include -I/tmp/neovim20151026-17981-bt0ppn/deps-build/usr/include -I/tmp/neovim20151026-17981-bt0ppn/deps-build/usr/include -I/tmp/neovim20151026-17981-bt0ppn/deps-build/usr/include -I/home/comp/chengxu/usr/include -I/tmp/neovim20151026-17981-bt0ppn/build/src/nvim/auto -I/tmp/neovim20151026-17981-bt0ppn/build/include
Compiled by chengxu@csr40.comp.hkbu.edu.hk

Optional features included (+) or not (-): +acl   +iconv    +jemalloc
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/nvimrc"
     user vimrc file: "~/.nvimrc"
 2nd user vimrc file: "~/.nvim/nvimrc"
      user exrc file: "~/.exrc"
  fall-back for $VIM: "/home/comp/chengxu/usr/Cellar/neovim/HEAD/share/nvim"
```

justinmk commented:
@xu-cheng It is intentional; back-compat make be broken in some cases before 1.0. Here the migration is minimal, see [`:help nvim-from-vim`](https://github.com/ZyX-I/neovim/blob/42047acb4f07c689936b051864c6b4448b1b6aa1/runtime/doc/nvim_from_vim.txt#L12-L18) for steps.

tarruda commented:
Great, will update my nvim version with this branch. Thanks to all involved in this.

BTW could #3498 be caused by this?

wbinnssmith commented:
hey folks,

anyone know where the default location for the `autoload` directory went? neovim's not picking up up (for example, [`plug.vim`](https://github.com/junegunn/vim-plug) in $XDG_DATA_DIR or $XDG_CONFIG_DIR.

`:help autoload`, points to `runtimepath`, and `:help runtimepath` claims that these directories are evaluated, but I can't reproduce this.

fmoralesc commented:
@wbinnssmith In that case, you should place the autoload directory in `~/.config/nvim/` or whatever `split(runtimepath, ',', 0)[0]` is. It should be `$XDG_CONFIG_DIR/nvim/`.

ZyX-I commented:
@wbinnssmith Location of autoload directory did not change *at all*. The only thing which *did* change is the default value of &runtimepath. You may create `autoload` subdirectory in any directory you will see in `:set runtimepath?` output.

ZyX-I commented:
@wbinnssmith If you are not using some plugin manager it is now expected that you put `plug.vim` (i.e. third-party autoload file) into $XDG_DATA_HOME/nvim/autoload. `$XDG_DATA_HOME` defaults to `~/.local/share`.

Note: putting it into $XDG_CONFIG_HOME (which is ~/.config) by default will also work, but this is going against XDG spec: `plug.vim` is not a user configuration file.

ZyX-I commented:
@fmoralesc There is no `$XDG_CONFIG_DIR`: there are variables `$XDG_CONFIG_HOME` (one directory, default `~/.config`) and `$XDG_CONFIG_DIRS` (colon-separated list, default `/etc/xdg`).

wbinnssmith commented:
> Location of autoload directory did not change at all. The only thing which did change is the default value of &runtimepath. You may create autoload subdirectory in any directory you will see in :set runtimepath? output.

This is exactly what I expected it would be. Thanks :smile: 

re: CONFIG_DIR, that was my slipup here on the comments.

> @wbinnssmith If you are not using some plugin manager it is now expected that you put plug.vim (i.e. third-party autoload file) into $XDG_DATA_HOME/nvim/autoload. $XDG_DATA_HOME defaults to ~/.local/share.

I'm actually *trying* to use a plugin manager in this case, which would in this case be the only file in `autoload`, yes? I agree that `$XDG_DATA_HOME` is a better place for this over CONFIG, but unfortunately neither are panning out for me at the moment. I'll continue looking into it.

fmoralesc commented:
@ZyX-I Sorry, you are right about `XDG_CONFIG_DIR`; dunno where I got that from.

It's worth noting that `~/.local/share/nvim/autoload` won't work: the default &runtimepath does not use that but `~/.local/share/nvim/site`. (Why is that, btw?)

ZyX-I commented:
@fmoralesc You are right, I forgot about `site`. About why: default for `$XDG_DATA_DIRS` is `/usr/local/share/:/usr/share/`. `$VIMRUNTIME` is normally `/usr/share/nvim/runtime`. So I used `$xdg_data_dir/nvim/site` to make structure similar to the current Vim variant: `$xdg_data_dir/vim/vim74` ($VIMRUNTIME) + `$xdg_data_dir/vim/vimfiles` (third-party files). $XDG_DATA_HOME got `site` mainly because I have choosen to treat it identically to directories in `$XDG_DATA_DIRS`.

Additionally users that install Neovim in $HOME may use the same layout: `VIMRUNTIME=$HOME/.local/share/nvim/runtime` with plugins in `$HOME/.local/share/nvim/site`.

@wbinnssmith Note that I forgot about `site` subdirectory: `~/.local/share/nvim/site/autoload`, not `~/.local/share/nvim/autoload`. Though this should be seen in `set rtp?`.

wbinnssmith commented:
@ZyX-I yeah, I resolved my issue — it was unrelated to the `autoload` path and just coincided with moving my configuration to XDG :smile: 

I do think that `~/.local/share/nvim` (`$XDG_DATA_HOME`) should be in the runtimepath though. Do you agree?

ZyX-I commented:
@wbinnssmith No. I have said why `site` subdirectory is there and it is useful.
